### PR TITLE
feat(balance, port): arpen reduction via layered armor, make shields block ranged hits

### PR DESCRIPTION
**Pull request by chaosvolt**
## Summary
SUMMARY: Balance "Fix shields not blocking ranged hits anymore, port over arpen being reduced by multi-layered armor from DDA"

## Purpose of change

This addresses an issue that was pointed out a while back with the ammo rebalance affects, namely that a high armor penetration is good and working as intended against a single high-armor item, but undermines the effectiveness of layering low-mid value items. Or as the PR OP put it:

> Armor penetration is broken, both because it has absurdly high values and also because it renders single armor layers useless (you could wear 20 individual hypothetical 1mm kevlar shirts and still take full damage from a 5.56 nato round, which has 6 armor penetration, but if you had a single 8mm thick kevlar shirt, it would stop the bullet almost entirely)

This is actually more relevant in BN than in DDA because 6 arpen is like nothing in BN compared to what rifle rounds (especially AP rounds, like 5.56 which gets 32) tend to have. This also makes shields, which have a chance to intercept ranged attacks, more consistently useful against bullets since instead of doing literally nothing to a shot that punches though, it'll at least make it a bit easier on the underlying armor.

Along the way, while testing I discovered a bug caused by https://github.com/cataclysmbnteam/Cataclysm-BN/pull/2776 where basically block chance for ranged attacks is now ONLY being rolled for the OPPOSITE arm from what it's worn on.

## Describe the solution

1. In item.cpp, ported over a change to `item::mitigate_damage` that reduces arpen by how much damage it broke through.
2. In the same function, per Kheir's advice tweaked the way the damage mitigation part is handled to tack on its `std:max` a bit more elegantly.
3. In character.cpp, also implemented the same general change to `Character::block_ranged_hit` so that blocking a ranged attack with a shield will affect arpen the same as it would if it hit the body parts it covers.
4. In the same function, fixed ranged blocking by setting it to use `get_avg_coverage` instead of `get_coverage` to determine block chance.

## Describe alternatives you've considered

If anyone would rather I split off the fixing of blocking ranged hits with shields into a separate PR, lemme know.

## Testing

First, tested pre-PR behavior in build 2023-10-24-1439 by layering on two light survivor suits (14 ballistic protection) and spawning in a light turret (uses 9x19mm FMJ, base damage 27, base arpen 15). Damage taken tended to be 28-32.

Second test in said build, spawned in a riot shield (8 ballistic protection) and wore that over a single heavy survivor suit (26 ballistic resistance). Damage taken on a sucessful range block tended to be 18-20. Damage rolls only triggered when the opposite arm was hit.

1. Compiled and load-tested.
2. Redid first test with a pair of light survivor suits vs a light turret.
3. Damage tended to be in the 16-18 range now.
5. Repeated second test with a riot shield and heavy survivor suit.
6. Damaged tended to be 11-12 for hits that punched through the shield and hit the survivor suit.
7. Also observed that hits to the arm it's worn on tended to be in the same 11-12 damage range as blocked shots, so the riot shield is doing the same thing as both armor and when blocking ranged hits to other parts.
8. Confirmed as well via debug that block roll is back to correctly being 90% chance to block when hitting normal parts, and 75% for leg hits.
9. Checked affected files for astyle.

## Additional context

DDA PR, albative arpen effects by @anoobindisguise: https://github.com/CleverRaven/Cataclysm-DDA/pull/66653

### CODE REVIEW

**Review by scarf005 — DISMISSED** [review]
i don't like that `Character::block_ranged_hit` and `item::mitigate_damage` has to duplicate logic (double bug potential!) but forgive me for my sin

**Comment by chaosvolt** [comment]
Same, perhaps it can in a followup be set to call mitigate damage directly? Not sure how it works so.

**Comment by RoyalFox2140** [comment]
Can we get some tests done on NPC's to make sure the armor/damage is correctly applied to them? I have balance changes in mind to mods if it applies correctly on all use cases.

**Comment by scarf005** [comment]
## Before

image

## After

image

image

certainly wearing more clothings reduce total damage taken.

**Comment by chaosvolt** [comment]
> Can we get some tests done on NPC's to make sure the armor/damage is correctly applied to them? I have balance changes in mind to mods if it applies correctly on all use cases.

Shooting an NPC in the face with two survivor hoods stacked on the head, before (M4):
image

After:
image


UHHHH

**Comment by scarf005** [comment]
> i don't like that `Character::block_ranged_hit` and `item::mitigate_damage` has to duplicate logic (double bug potential!) but forgive me for my sin

i suspect this is why
tl;dr duplicate logics are duplicated

**Comment by chaosvolt** [comment]
```
void item::mitigate_damage( damage_unit &du ) const
{
    add_msg( m_warning, _( "If this message prints, whoever just got fucking capped correctly called item::mitigate_damage." ) );
```
image

So thing is, the function that should be working right now is the one that's supposed to be for armor. So the issue isn't the sin of duplication we're committing here, it's that someone else has already sinned and is evidently using an entirely different function for NPC armor effects.

**Comment by chaosvolt** [comment]
Further discoveries from the BN discord:
1. NPCs call `Character::absorb_hit` as expected, but not `item::mitigate_damage` like it does when the PLAYER gets hit.
2. Players also seem to spam the fuck out of `monster::absorb_hit` on their turns?

image
image
image

**Comment by chaosvolt** [comment]
So I decided to repeat my "9x19mm FMJ vs double survivor suit" test and it...works?
image

My mistake was that "5.56mm vs. two survivor suits" is just so much arpen that I forgot that the damage would still be unchanged.

File: src/item.cpp [hunk 1]
```diff
@@ -6194,6 +6194,10 @@ void item::mitigate_damage( damage_unit &du ) const
 {
     const resistances res = resistances( *this );
     const float mitigation = res.get_effective_resist( du );
+    // get_effective_resist subtracts the flat penetration value before multiplying the remaining armor.
+    // therefore, res_pen is reduced by the full value of the item's armor value even though mitigation might be smaller (such as an attack with a 0.5 armor multiplier)
+    du.res_pen -= res.type_resist( du.type );
+    du.res_pen = std::max( 0.0f, du.res_pen );
```
Comment (KheirFerrum) [inline src/item.cpp:6200]: Could do something like
`du.res_pen = std::max( 0.0f, du.res_pen - res.type_resist( du.type );`

**Review by scarf005 — APPROVED** [review]
